test/unit/state_resolver_test.rb

```diff
@@ -0,0 +1,171 @@
+            when "no" then question :a
+            end
+          end
+        end
+
+        radio :y do
+          option :yes
+          option :no
+          next_node do |response|
+            case response
+            when "yes" then outcome :a
+            when "no" then outcome :b
+            end
+          end
+        end
+        outcome :a
+        outcome :b
+      end
+
+      @state_resolver = StateResolver.new(@flow)
+    end
+
+    context "#state_from_response_store" do
+      should "return the state for the first question when there are no responses" do
+        response_store = ResponseStore.new(responses: {})
+        state = @state_resolver.state_from_response_store(response_store)
+
+        assert_equal :x, state.current_node
+        assert_empty state.accepted_responses
+        assert_nil state.current_response
+      end
+
+      should "return the state for the next question when there are valid answers" do
+        response_store = ResponseStore.new(responses: { "x" => "yes" })
+        state = @state_resolver.state_from_response_store(response_store)
+
+        assert_equal :y, state.current_node
+        assert_equal ({ x: "yes" }), state.accepted_responses
+        assert_nil state.current_response
+      end
+
+      should "return an error state for the current question when an answer is invalid" do
+        response_store = ResponseStore.new(responses: { "x" => "yes", "y" => "invalid" })
+        state = @state_resolver.state_from_response_store(response_store)
+
+        assert_equal :y, state.current_node
+        assert_equal ({ x: "yes" }), state.accepted_responses
+        assert_equal "invalid", state.current_response
+        assert state.error.present?
+      end
+
+      should "return an outcome state when all questions are answered" do
+        response_store = ResponseStore.new(responses: { "x" => "yes", "y" => "no" })
+        state = @state_resolver.state_from_response_store(response_store)
+
+        assert_equal :b, state.current_node
+        assert_equal ({ x: "yes", y: "no" }), state.accepted_responses
+        assert_nil state.current_response
+      end
+
+      should "return an particular node if one is requested" do
+        response_store = ResponseStore.new(responses: { "x" => "yes", "y" => "no" })
+        state = @state_resolver.state_from_response_store(response_store, "y")
+
+        assert_equal :y, state.current_node
+        assert_equal ({ x: "yes" }), state.accepted_responses
+        assert_equal "no", state.current_response
+      end
+
+      should "return an error state for an earlier node regardless of a requested node" do
+        response_store = ResponseStore.new(responses: { "x" => "invalid", "y" => "no" })
+        state = @state_resolver.state_from_response_store(response_store, "y")
+
+        assert_equal :x, state.current_node
+        assert_equal ({}), state.accepted_responses
+        assert_equal "invalid", state.current_response
+      end
```

Comment:
Think it'd be good to include a test case where there's an invalid response for the requested node - so we check that the state actually captures errors

```ruby
should "return an error state for invalid response on requested node" do
  response_store = ResponseStore.new(responses: { "x" => "valid", "y" => "invalid" })
  state = @state_resolver.state_from_response_store(response_store, "y")

  assert_equal :y, state.current_node
  assert_equal ({x:  "valid"}), state.accepted_responses
  assert_equal "invalid", state.current_response
  assert state.error.present?
end
```

This would catch a issue where state is returned before having attempted to transition and checking if an error exists: E.g.
```ruby
    def state_from_response_store(response_store, requested_node = nil)
       forwarding_responses = @flow.response_store == :session ? {} : response_store.all
       state = start_state(forwarding_responses)
       loop do
         node_name = state.current_node.to_s
         return state unless response_store.has?(node_name)

         response = response_store.get(node_name)

->       # return apply_response_to_state(state, response) if node_name == requested_node
->       return state if node_name == requested_node
         
         new_state = transition_state_to_next_node(state, response)
         return new_state if new_state.error
         return new_state if @flow.node(new_state.current_node).outcome?

         state = new_state
       end
     end
```

Reply:
Sure I've added this and your other requested test. I don't find myself convinced by their necessity, but I can understand why you've requested them, and decided I'm not super concerned. 

I do think though that they're risking crossing from testing the unit of code into testing how a developer could imagine the code breaking were someone to change it, since they're testing code that is already tested in a different way. This seems a more ambiguous approach to deciding what to test and I find it can leave a number of "what ifs" as to why we're testing some scenarios and not other imaginable ones.

Reply:
Thanks! Sorry I think I wrote the comment as if I was concerned with the implementation details. But I was thinking about in terms of behaviour we expect this function to have. We definitely want to make sure that if a user requests a node and that node has an invalid previous response, we want to make sure we show that error. Just added the code example to illustrate how the behaviour might fail. (It's also a bug we've had in the past.)